test/e2e/utils.go

```diff
@@ -0,0 +1,210 @@
+import (
+	goctx "context"
+	"github.com/integr8ly/keycloak-operator/pkg/apis/aerogear/v1alpha1"
+	appsv1 "github.com/openshift/client-go/apps/clientset/versioned/typed/apps/v1"
+	framework "github.com/operator-framework/operator-sdk/pkg/test"
+	"github.com/operator-framework/operator-sdk/pkg/test/e2eutil"
+	"github.com/sirupsen/logrus"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/rest"
+	"testing"
+	"time"
+)
+
+func prepare(t *testing.T) *framework.TestCtx {
+	ctx := framework.NewTestCtx(t)
+
+	opt := &framework.CleanupOptions{
+		TestContext:   ctx,
+		RetryInterval: retryInterval,
+		Timeout:       timeout,
+	}
+
+	err := ctx.InitializeClusterResources(opt)
+	if err != nil {
+		t.Fatalf("Failed to initialize test context: %v", err)
+	}
+
+	ns, err := ctx.GetNamespace()
+	if err != nil {
+		t.Fatalf("Failed to get context namespace: %v", err)
+	}
+
+	globalVars := framework.Global
+
+	err = e2eutil.WaitForDeployment(t, globalVars.KubeClient, ns, "keycloak-operator", 1, retryInterval, timeout)
+	if err != nil {
+		t.Fatalf("Operator deployment failed: %v", err)
+	}
+
+	return ctx
+}
+
+func register() error {
+	var err error
+
+	keycloakList := &v1alpha1.KeycloakList{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       v1alpha1.KeycloakKind,
+			APIVersion: v1alpha1.Group + "/" + v1alpha1.Version,
+		},
+	}
+	err = framework.AddToFrameworkScheme(v1alpha1.AddToScheme, keycloakList)
+	if err != nil {
+		return err
+	}
+
+	keycloak := &v1alpha1.Keycloak{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       v1alpha1.KeycloakKind,
+			APIVersion: v1alpha1.Group + "/" + v1alpha1.Version,
+		},
+	}
+	err = framework.AddToFrameworkScheme(v1alpha1.AddToScheme, keycloak)
+	if err != nil {
+		return err
+	}
+
+	keycloakRealmList := &v1alpha1.KeycloakRealmList{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       v1alpha1.KeycloakRealmKind,
+			APIVersion: v1alpha1.Group + "/" + v1alpha1.Version,
+		},
+	}
+	err = framework.AddToFrameworkScheme(v1alpha1.AddToScheme, keycloakRealmList)
+	if err != nil {
+		return err
+	}
+
+	keycloakRealm := &v1alpha1.KeycloakRealm{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       v1alpha1.KeycloakRealmKind,
+			APIVersion: v1alpha1.Group + "/" + v1alpha1.Version,
+		},
+	}
+	err = framework.AddToFrameworkScheme(v1alpha1.AddToScheme, keycloakRealm)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func cleanupOpts(ctx *framework.TestCtx) *framework.CleanupOptions {
+	return &framework.CleanupOptions{
+		TestContext:   ctx,
+		Timeout:       timeout,
+		RetryInterval: retryInterval,
+	}
+}
+
+func doDeployment(f *framework.Framework, ctx *framework.TestCtx, keycloakCr *v1alpha1.Keycloak, realmCr *v1alpha1.KeycloakRealm) error {
+
+	err := f.Client.Create(goctx.TODO(), realmCr, cleanupOpts(ctx))
+	if err != nil {
+		return err
+	}
+
+	err = f.Client.Create(goctx.TODO(), keycloakCr, cleanupOpts(ctx))
```

Comment:
why do we create the realm before the keycloak, does it matter?

Reply:
the order doesn't really matter but we need both CRs for the deployment to happen.